File: olim/api/schemas/pipeline.py
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from olim.pipelines import BlockType


class _BlockInBase(BaseModel):
    model_config = ConfigDict(extra="forbid")


class LowercaseBlockIn(_BlockInBase):
    type: Literal["lowercase"]


class StripStopwordsBlockIn(_BlockInBase):
    type: Literal["strip_stopwords"]


class TfidfBlockIn(_BlockInBase):
    type: Literal["tfidf"]
    max_features: Annotated[int, Field(gt=0)] | None = None
    ngram_max: Annotated[int, Field(ge=1)] = 1


class CountVectorizerBlockIn(_BlockInBase):
    type: Literal["count_vectorizer"]
    max_features: Annotated[int, Field(gt=0)] | None = None
    ngram_max: Annotated[int, Field(ge=1)] = 1


class TrainTestSplitBlockIn(_BlockInBase):
    type: Literal["train_test_split"]
    test_size: Annotated[float, Field(gt=0, lt=1)] = 0.2


class TrainCalibTestSplitBlockIn(_BlockInBase):
    type: Literal["train_calib_test_split"]
    test_size: Annotated[float, Field(gt=0, lt=1)] = 0.2
    calib_size: Annotated[float, Field(gt=0, lt=1)] = 0.2


class LogRegBlockIn(_BlockInBase):
    type: Literal["logreg"]
    C: Annotated[float, Field(gt=0)] = 1.0


class RandomForestBlockIn(_BlockInBase):
    type: Literal["random_forest"]
    n_estimators: Annotated[int, Field(gt=0)] = 100


class XGBoostBlockIn(_BlockInBase):
    type: Literal["xgboost"]
    n_estimators: Annotated[int, Field(gt=0)] = 100
    learning_rate: Annotated[float, Field(gt=0)] = 0.3


class ConformalBlockIn(_BlockInBase):
    type: Literal["conformal"]
    alpha: Annotated[float, Field(gt=0, lt=1)] = 0.1


class ClassificationMetricsBlockIn(_BlockInBase):
    type: Literal["classification_metrics"]


class PickleArtifactBlockIn(_BlockInBase):
    type: Literal["pickle_artifact"]


BlockIn = Annotated[
    LowercaseBlockIn
    | StripStopwordsBlockIn
    | TfidfBlockIn
    | CountVectorizerBlockIn
    | TrainTestSplitBlockIn
    | TrainCalibTestSplitBlockIn
    | LogRegBlockIn
    | RandomForestBlockIn
    | XGBoostBlockIn
    | ConformalBlockIn
    | ClassificationMetricsBlockIn
    | PickleArtifactBlockIn,
    Field(discriminator="type"),
]

_CONFIG_MODELS: dict[BlockType, type[_BlockInBase]] = {
    "lowercase": LowercaseBlockIn,
    "strip_stopwords": StripStopwordsBlockIn,
    "tfidf": TfidfBlockIn,
    "count_vectorizer": CountVectorizerBlockIn,
    "train_test_split": TrainTestSplitBlockIn,
    "train_calib_test_split": TrainCalibTestSplitBlockIn,
    "logreg": LogRegBlockIn,
    "random_forest": RandomForestBlockIn,
    "xgboost": XGBoostBlockIn,
    "conformal": ConformalBlockIn,
    "classification_metrics": ClassificationMetricsBlockIn,
    "pickle_artifact": PickleArtifactBlockIn,
}


def config_schema_for(block_type: BlockType) -> dict:
    return _CONFIG_MODELS[block_type].model_json_schema()


class PipelineCreateIn(BaseModel):
    name: str


class CandidateOut(BaseModel):
    """A block that may be appended next, with its config schema for the form."""

    type: BlockType
    category: str
    consumes: list[str]
    produces: list[str]
    config_schema: dict

File: web-client/app/types/pipeline.ts
import type { BlockType } from "./common";

export interface PipelineBlockDTO {
  id: number;
  pipeline_id: number;
  type: BlockType;
  position: number;
  config: Record<string, unknown>;
}

export interface PipelineDTO {
  id: number;
  dataset_id: number;
  scheme_id: number;
  name: string;
  blocks: PipelineBlockDTO[];
}

export interface PipelineCreateIn {
  name: string;
}

export interface CandidateOut {
  type: BlockType;
  category: string;
  consumes: string[];
  produces: string[];
  config_schema: Record<string, unknown>;
}

// Block configs accepted by POST /pipelines/{id}/blocks (discriminated on `type`).
export interface LowercaseBlockIn {
  type: "lowercase";
}
export interface StripStopwordsBlockIn {
  type: "strip_stopwords";
}
export interface TfidfBlockIn {
  type: "tfidf";
  max_features?: number | null;
  ngram_max?: number;
}
export interface CountVectorizerBlockIn {
  type: "count_vectorizer";
  max_features?: number | null;
  ngram_max?: number;
}
export interface TrainTestSplitBlockIn {
  type: "train_test_split";
  test_size?: number;
}
export interface TrainCalibTestSplitBlockIn {
  type: "train_calib_test_split";
  test_size?: number;
  calib_size?: number;
}
export interface LogRegBlockIn {
  type: "logreg";
  C?: number;
}
export interface RandomForestBlockIn {
  type: "random_forest";
  n_estimators?: number;
}
export interface XGBoostBlockIn {
  type: "xgboost";
  n_estimators?: number;
  learning_rate?: number;
}
export interface ConformalBlockIn {
  type: "conformal";
  alpha?: number;
}
export interface ClassificationMetricsBlockIn {
  type: "classification_metrics";
}
export interface PickleArtifactBlockIn {
  type: "pickle_artifact";
}

export type BlockIn =
  | LowercaseBlockIn
  | StripStopwordsBlockIn
  | TfidfBlockIn
  | CountVectorizerBlockIn
  | TrainTestSplitBlockIn
  | TrainCalibTestSplitBlockIn
  | LogRegBlockIn
  | RandomForestBlockIn
  | XGBoostBlockIn
  | ConformalBlockIn
  | ClassificationMetricsBlockIn
  | PickleArtifactBlockIn;
